ports/esp32s2/common-hal/busio/SPI.c

```diff
@@ -0,0 +1,358 @@
+        host_id = SPI2_HOST;
+        self->connected_through_gpio = false;
+    } else if (spi_bus_is_free(SPI3_HOST)) {
+        host_id = SPI3_HOST;
+    } else if (spi_bus_is_free(SPI2_HOST)) {
+        host_id = SPI2_HOST;
+    }
+    if (host_id == SPI1_HOST) {
+        mp_raise_ValueError(translate("All SPI peripherals are in use"));
+    }
+
+    esp_err_t result = spi_bus_initialize(host_id, &bus_config, host_id /* dma channel */);
+    if (result == ESP_ERR_NO_MEM) {
+        mp_raise_msg(&mp_type_MemoryError, translate("ESP-IDF memory allocation failed"));
+    } else if (result == ESP_ERR_INVALID_ARG) {
+        mp_raise_ValueError(translate("Invalid pins"));
+    }
+
+    // After this point, we need to deinit to free IDF memory so fill out self's pins.
+    self->clock_pin = clock;
+    self->MOSI_pin = mosi;
+    self->MISO_pin = miso;
+    self->host_id = host_id;
+
+    spi_bus_lock_dev_config_t config = { .flags = 0 };
+    // The returned lock is stored in the bus lock but must be freed separately with
+    // spi_bus_lock_unregister_dev.
+    result = spi_bus_lock_register_dev(spi_bus_get_attr(host_id)->lock,
+                                       &config,
+                                       &self->lock);
+    if (result == ESP_ERR_NO_MEM) {
+        common_hal_busio_spi_deinit(self);
+        mp_raise_msg(&mp_type_MemoryError, translate("ESP-IDF memory allocation failed"));
+    }
+    lock_dev_handle[host_id] = self->lock;
+
+    result = esp_intr_alloc(spicommon_irqsource_for_host(host_id),
+                            bus_config.intr_flags | ESP_INTR_FLAG_INTRDISABLED,
+                            spi_interrupt_handler, self, &self->interrupt);
+    if (result == ESP_ERR_NO_MEM) {
+        common_hal_busio_spi_deinit(self);
+        mp_raise_msg(&mp_type_MemoryError, translate("ESP-IDF memory allocation failed"));
+    }
+    intr_handle[host_id] = self->interrupt;
+    spi_bus_lock_set_bg_control(spi_bus_get_attr(host_id)->lock, spi_bus_intr_enable, spi_bus_intr_disable, self);
+
+    spi_hal_context_t* hal = &self->hal_context;
+
+    // spi_hal_init clears the given hal context so set everything after.
+    spi_hal_init(hal, host_id);
+    hal->dmadesc_tx = &self->tx_dma;
+    hal->dmadesc_rx = &self->rx_dma;
+    hal->dmadesc_n = 1;
+
+    // We don't use native CS.
+    // hal->cs_setup = 0;
+    // hal->cs_hold = 0;
+    // hal->cs_pin_id = 0;
+
+    hal->sio = 0;
+    hal->half_duplex = 0;
+    // hal->tx_lsbfirst = 0;
+    // hal->rx_lsbfirst = 0;
+    hal->dma_enabled = 1;
+    hal->no_compensate = 1;
+    // Ignore CS bits
+
+    // We don't use cmd, addr or dummy bits.
+    // hal->cmd = 0;
+    // hal->cmd_bits = 0;
+    // hal->addr_bits = 0;
+    // hal->dummy_bits = 0;
+    // hal->addr = 0;
+
+    hal->io_mode = SPI_LL_IO_MODE_NORMAL;
+
+    // This must be set after spi_hal_init.
+    hal->timing_conf = &self->timing_conf;
```

Comment:
Another internal pointer, how is this going to play with moving to the long-lived heap?

Reply:
Ok, I've fixed this by setting it twice before it is used. It isn't held by the IDF.